Compiler errors and unit test failure in `IOPool/Common` for el9_x_gcc11

Hello,

There are some both compiler errors and a unit test failure in package `IOPool/Common` in last night IBs [CMSSW_12_5_X_2022-07-28-2300](https://cmssdt.cern.ch/SDT/cgi-bin/showBuildLogs.py/el9_amd64_gcc11/www/thu/12.5-thu-23/CMSSW_12_5_X_2022-07-28-2300?utests) for `el9_amd64_gcc11` and `el9_aarch64_gcc11`.

- Compiler errors:
```
>> Compiling  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc
/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/gcc/11.2.1-f478fee2760dbd22aaabb4e3a8fe1640/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=110201 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_12_5_X_2022-07-28-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_12_5_X_2022-07-28-2300' -I/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src -I/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/poison -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/cms/cmssw/CMSSW_12_5_X_2022-07-24-2300/src -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/pcre/8.43-e6b3d6f8e1424033990367f0ae94d8f3/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/boost/1.78.0-9be29a07a3b36f8166c42646282ae8f6/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/bz2lib/1.0.6-a8795c5492b25410af2fdbfc34f86eab/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/clhep/2.4.5.1-863959129834eec6fab65756f04dfddf/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/libuuid/2.34-548af62999d8f41d39f61053ab3a9206/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/lcg/root/6.24.07-71d7866f2340bdbbcee492f72c476827/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/tbb/v2021.5.0-e991d20102976c9e7ac50996a50a8ae4/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/xz/5.2.5-4fdf584c0d81808d2831b2beffcb3bc7/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/zlib/1.2.11-31d2ceb65830b90a6cc7a5a90279f5d6/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/fmt/8.0.1-2a6682ea4a36cb34727e1524caeeffb8/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/md5/1.0.0-65713d22725e6dea36af6865a8b7982a/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/tinyxml2/6.2.0-dff3d5cfeccbaf05917bf8c158421d93/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++1z -ftree-vectorize -Wstrict-overflow -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS  -fPIC  -MMD -MF tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/EdmProvDump.cc.d /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc -o tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/EdmProvDump.cc.o
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc:495:8: error: 'optional' in namespace 'std' does not name a template type
   495 |   std::optional<std::tuple<edm::BranchIDListHelper, std::vector<edm::ProcessIndex>>> makeBranchIDListHelper();
      |        ^~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc:32:1: note: 'std::optional' is defined in header '<optional>'; did you forget to '#include <optional>'?
   31 | #include <iostream>
  +++ |+#include <optional>
   32 | #include <memory>
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc:637:6: error: 'optional' in namespace 'std' does not name a template type
   637 | std::optional<std::tuple<edm::BranchIDListHelper, std::vector<edm::ProcessIndex>>>
      |      ^~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc:637:1: note: 'std::optional' is defined in header '<optional>'; did you forget to '#include <optional>'?
  637 | std::optional<std::tuple<edm::BranchIDListHelper, std::vector<edm::ProcessIndex>>>
      | ^~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc: In member function 'void ProvenanceDumper::work_()':
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/src/IOPool/Common/bin/EdmProvDump.cc:793:46: error: 'makeBranchIDListHelper' was not declared in this scope
   793 |   auto branchIDListHelperAndToProcessIndex = makeBranchIDListHelper();
      |                                              ^~~~~~~~~~~~~~~~~~~~~~
  gmake: *** [tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/EdmProvDump.cc.o] Error 1
 >> Building binary edmProvDump
/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/gcc/11.2.1-f478fee2760dbd22aaabb4e3a8fe1640/bin/c++ -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++1z -ftree-vectorize -Wstrict-overflow -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -fPIC tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/EdmProvDump.cc.o -Wl,-E -Wl,--hash-style=gnu -L/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/biglib/el9_amd64_gcc11 -L/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/lib/el9_amd64_gcc11 -L/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/0a3e5311dbfc304dafc6cb454a565aab/opt/cmssw/el9_amd64_gcc11/cms/cmssw-patch/CMSSW_12_5_X_2022-07-28-2300/external/el9_amd64_gcc11/lib -L/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/cms/cmssw/CMSSW_12_5_X_2022-07-24-2300/lib/el9_amd64_gcc11 -L/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/cms/cmssw/CMSSW_12_5_X_2022-07-24-2300/biglib/el9_amd64_gcc11 -lFWCoreServices -lFWCoreCatalog -lFWCoreServiceRegistry -lSimDataFormatsRandomEngine -lDataFormatsCommon -lFWCoreParameterSet -lFWCoreMessageLogger -lDataFormatsProvenance -lFWCorePluginManager -lFWCoreReflection -lFWCoreUtilities -lHist -lMatrix -lTree -lNet -lThread -lMathCore -lRIO -lboost_program_options -lCore -lboost_thread -lboost_date_time -lCLHEP -lpcre -lbz2 -luuid -ltbb -llzma -lz -lfmt -lcms-md5 -lcrypt -ldl -lrt -lstdc++fs -ltinyxml2 -o tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/edmProvDump
/data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc11/external/gcc/11.2.1-f478fee2760dbd22aaabb4e3a8fe1640/bin/../lib/gcc/x86_64-redhat-linux-gnu/11.2.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/EdmProvDump.cc.o: No such file or directory
collect2: error: ld returned 1 exit status
>> Deleted: tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/edmProvDump
  gmake: *** [tmp/el9_amd64_gcc11/src/IOPool/Common/bin/edmProvDump/edmProvDump] Error 1
 Entering library rule at IOPool/Common
 ```

- Unit test failure in test `TestEdmConfigDump`:
```
---
>   edmtestIntProduct_source__PROD1.
Failure comparing provdump.log: status 1
status = 256
```

Could it be caused by https://github.com/cms-sw/cmssw/pull/38858/?

Many thanks,
Andrea.

--- a/IOPool/Common/bin/EdmProvDump.cc
+++ b/IOPool/Common/bin/EdmProvDump.cc
@@ -31,6 +31,7 @@
 #include <iostream>
 #include <memory>
 #include <map>
+#include <optional>
 #include <set>
 #include <sstream>
 #include <vector>

